BHBC-1594: Upload Documents / Files to a Project

shreypdev opened the pull request:
# Overview

## Links to jira tickets

- [BHBC-1594](https://quartech.atlassian.net/browse/BHBC-1594)

## This PR contains the following changes

- Authenticated user can upload, view and delete the document.
- Public user can view the document that are uploaded.

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

![image](https://user-images.githubusercontent.com/60997794/155595823-987951d8-f98a-4c20-8038-72dfa8419295.png)

![image](https://user-images.githubusercontent.com/60997794/155595892-2e369650-9c36-43a5-8771-9fc6efd5a219.png)

shreypdev committed:
Changed API according to the requirements

shreypdev committed:
Fixed and added all API broken test cases

shreypdev committed:
Api minor changes

shreypdev committed:
Made changes to UI and fixed the broken FE tests

NickPhura commented on api/src/services/attachment-service.ts:
```diff
@@ -0,0 +1,91 @@
+  async insertProjectAttachment(projectId: number, key: string, file: Express.Multer.File): Promise<{ id: number }> {
+    const sqlStatement = queries.project.postProjectAttachmentSQL(file.originalname, file.size, projectId, key);
+
+    if (!sqlStatement) throw new HTTP400('Failed to build SQL insert statement');
```
Can we add the brackets here (and wherever else). Its generally better, as far as I've read, to have them even for 1 liners just to help with legibility and reduce the chance of errors by maintainers in future.

shreypdev replied:
Sure thing.

shreypdev committed:
Fixed the code styles

anissa-agahchen reviewed (changes requested):
@shreypdev the typical behaviour in SIMS is to overwrite a file is the file name is the same. 
![Screen Shot 2022-02-24 at 2 35 47 PM](https://user-images.githubusercontent.com/22803003/155619088-caa22ec9-2d46-4767-be67-cb4580b69c07.png)

shreypdev committed:
Fixed issue of same file name by replacing existing one

anissa-agahchen commented on api/src/paths/project/{projectId}/attachments/list.test.ts:
```diff
@@ -38,141 +37,11 @@ describe('lists the project attachments', () => {
     sinon.restore();
   });
 
-  it('should throw a 400 error when no sql statement returned for getProjectAttachmentsSQL', async () => {
```
why remove this test?

shreypdev replied:
This test is moved to project service as no sql error is not thrown in list.test.ts, now it is thrown by attachment-service.

anissa-agahchen commented on api/src/paths/project/{projectId}/attachments/upload.ts:
```diff
@@ -142,72 +144,3 @@ export function uploadMedia(): RequestHandler {
     }
   };
 }
-
```
is this test not needed anymore?

shreypdev replied:
Sorry dont exactly know what you are pointing at here

anissa-agahchen commented on api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.test.ts:
```diff
@@ -83,128 +84,15 @@ describe('deleteAttachment', () => {
     }
   });
 
-  it('should throw a 400 error when no sql statement returned for unsecureAttachmentRecordSQL', async () => {
```
same question here .... not needed?

shreypdev replied:
Tests for unsecure is not needed as we no more switch between secure and unsecure. Also other delete tests are now located in attachment-service.

anissa-agahchen commented on api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.ts:
```diff
@@ -102,39 +95,3 @@ export function deleteAttachment(): RequestHandler {
   };
 }
-
-const unsecureProjectAttachmentRecord = async (securityToken: any, connection: IDBConnection): Promise<void> => {
```
not needed?

shreypdev replied:
No use of this according to the requirements

anissa-agahchen reviewed (commented):
@shreypdev the functionality is all good, just noticed that some tests were removed ... is that intentional?

shreypdev committed:
Minor Chnages
-> Added onClick event to cancel button for new project and edit project.
-> Removed expires limit from presigned S3 document link.
-> Code simplification in attachment service.

shreypdev committed:
Merge branch 'dev' into BHBC-1594

shreypdev committed:
format-fix
